Mathlib/Algebra/Order/SMul.lean

```diff
@@ -364,54 +364,55 @@ private theorem smul_ne_zero_of_ne_zero_of_pos [Preorder M] (ha : a ≠ 0) (hb :
---                   =>
---                   nonnegative <$> mk_app ` ` smul_nonneg_of_pos_of_nonneg [ pa , pb ]
---                 |
---                   nonnegative pa , positive pb
---                   =>
---                   nonnegative <$> mk_app ` ` smul_nonneg_of_nonneg_of_pos [ pa , pb ]
---                 |
---                   nonnegative pa , nonnegative pb
---                   =>
---                   nonnegative <$> mk_app ` ` smul_nonneg [ pa , pb ]
---                 |
---                   positive pa , nonzero pb
---                   =>
---                   nonzero <$> to_expr ` `( smul_ne_zero_of_pos_of_ne_zero $ ( pa ) $ ( pb ) )
---                 |
---                   nonzero pa , positive pb
---                   =>
---                   nonzero <$> to_expr ` `( smul_ne_zero_of_ne_zero_of_pos $ ( pa ) $ ( pb ) )
---                 |
---                   nonzero pa , nonzero pb
---                   =>
---                   nonzero <$> to_expr ` `( smul_ne_zero $ ( pa ) $ ( pb ) )
---                 | sa @ _ , sb @ _ => positivity_fail e a b sa sb
---       | e => pp e >>= fail ∘ format.bracket "The expression `" "` isn't of the form `a • b`"
--- #align tactic.positivity_smul Tactic.positivity_smul
-
-end Tactic
+open Lean.Meta Qq
+
+/-- Positivity extension for HSMul, i.e. (_ • _).  -/
+@[positivity HSMul.hSMul _ _]
+def evalHSMul : PositivityExt where eval {_u α} zα pα (e : Q($α)) := do
+  let .app (.app (.app (.app (.app (.app
+        (.const ``HSMul.hSMul [u1, _, _]) (M : Q(Type u1))) _) _) _)
+          (a : Q($M))) (b : Q($α)) ← whnfR e | throwError "failed to match hSMul"
+  let zM : Q(Zero $M) ← synthInstanceQ (q(Zero $M))
+  let pM : Q(PartialOrder $M) ← synthInstanceQ (q(PartialOrder $M))
+  match ← core zM pM a, ← core zα pα b with
+  | .positive pa, .positive pb =>
+      pure (.positive (← mkAppM ``smul_pos #[pa, pb]))
+  | .positive pa, .nonnegative pb =>
+      pure (.nonnegative (← mkAppM ``smul_nonneg_of_pos_of_nonneg #[pa, pb]))
+  | .nonnegative pa, .positive pb =>
+      pure (.nonnegative (← mkAppM ``smul_nonneg_of_nonneg_of_pos #[pa, pb]))
+  | .nonnegative pa, .nonnegative pb =>
+      pure (.nonnegative (← mkAppM ``smul_nonneg #[pa, pb]))
+  | .positive pa, .nonzero pb =>
+      pure (.nonzero (← mkAppM ``smul_ne_zero_of_pos_of_ne_zero #[pa, pb]))
+  | .nonzero pa, .positive pb =>
+      pure (.nonzero (← mkAppM ``smul_ne_zero_of_ne_zero_of_pos #[pa, pb]))
+  | .nonzero pa, .nonzero pb =>
+      pure (.nonzero (← mkAppM ``smul_ne_zero #[pa, pb]))
+  | _, _ => pure .none
+
+/-- Positivity extension for SMul. -/
+@[positivity SMul.smul _ _]
+def evalSMul : PositivityExt where eval {_u α} zα pα (e : Q($α)) := do
+  let .app (.app (.app (.app (.app
+        (.const ``SMul.smul [u1, _]) (M : Q(Type u1))) _) _)
+          (a : Q($M))) (b : Q($α)) ← whnfR e | throwError "failed to match smul"
+  -- TODO: deduplicate this logic, as it's shared with the HSMul extension, above.
```

Comment:
This shouldn't ever be needed: `SMul.smul` is an implementation detail.

Reply:
`evalAdd` is covers both `HAdd.hAdd` and `Add.add`:

https://github.com/leanprover-community/mathlib4/blob/00cb9455117f9b1046c7fa0d15bf5b8ad0654c1c/Mathlib/Tactic/Positivity/Basic.lean#L156

and `evalMul` covers both `HMul.mMul` and `Mul.mul`:
https://github.com/leanprover-community/mathlib4/blob/00cb9455117f9b1046c7fa0d15bf5b8ad0654c1c/Mathlib/Tactic/Positivity/Basic.lean#L197


Is there anything different about `SMul` that says we *don't* need to cover both cases here? (I would put `HSMul.hSMul` and `SMul.smul` in a single extension if I could figure out how.)

Reply:
I think all of these are mistakes, and could also be removed (in a separate PR)

Reply:
Removed the SMul extension in https://github.com/leanprover-community/mathlib4/pull/8067/commits/7876e072c6441fe1bfc33c0a309a13d8d95752b4